packages/axios/src/index.ts

```diff
@@ -1,52 +1,172 @@
+        err.request?.path
+      } status=${err.response?.status} body=${JSON.stringify(
+        err.response?.data
+      )}`
+    );
+    this.name = 'ExternalServiceError';
+  }
+}
+
+export default class Axios implements IModule, IAxios {
   name = 'axios';
-  private instances: Map<string, Axios> = new Map();
+  private instances: Map<string, A.Axios> = new Map();
+  private services: string[];
+  private nr: INewrelic | undefined;
+  constructor(services: string[]) {
+    this.services = services;
+  }
+
   async load(ctx: IContext) {
+    this.nr = ctx.provide<INewrelic>('newrelic');
+
     const config = ctx.provide<IConfig>('config');
     const axiosConfig = config?.get<IAxiosConfig>(this.name);
 
     if (!axiosConfig) {
       throw new Error(`Missing configuration for ${this.name}`);
     }
 
-    for (const [key, value] of Object.entries(axiosConfig.api)) {
+    for (const key of this.services) {
+      // Since we still don't have a static guarantee that service name `key` is an item of
+      // `axiosConfig.api`, throw if it's undefined.
+      if (!axiosConfig.api[key]) {
+        const sources = Config.files().join(', ');
+        throw new Error(
+          `Missing service definition "${key}" in all config sources: ${sources}`
+        );
+      }
+
       const instance = axios.create({
         ...(axiosConfig.global ?? {}),
-        ...value,
+        ...axiosConfig.api[key],
         httpsAgent: new https.Agent({ keepAlive: true }),
         httpAgent: new http.Agent({ keepAlive: true }),
       });
+
       this.instances.set(key, instance);
     }
   }
   async dispose() {
     this.instances.clear();
   }
 
-  api(name: string): Axios {
+  service(name: string, flags: RequestFlags = RequestFlags.None): TracedAxios {
+    const nr = this.nr;
+
     const instance = this.instances.get(name);
+
     if (!instance) {
-      throw new Error(`Missing requester for ${name}`);
+      throw new Error(
+        `Service ${name} is not configured. Available services: ${this.services.join(
+          ', '
+        )}`
+      );
     }
-    return instance;
+
+    // Else return a proxy that wraps the original instance
+    return new Proxy<TracedAxios>(instance as TracedAxios, {
+      get(target, prop, receiver) {
+        if (prop === 'trace')
+          return (ctx: ITrace) => {
+            target.defaults.headers.common[Headers.TRACE_ID] = ctx.trace.id;
+            return receiver;
+          };
+
+        // If not the method is not a HTTP method, return the original method
+        if (
+          ![
+            'request',
+            'get',
+            'post',
+            'put',
+            'delete',
+            'patch',
+            'head',
+            'options',
+            'postForm',
+            'putForm',
+            'patchForm',
+          ].includes(prop as string) ||
+          !nr
+        )
+          return Reflect.get(target, prop, receiver);
```

Comment:
Nit: I try to always leave style concerns to automated tools, but unbraced conditionals make me antsy. Does `prettier` enforce them when the body only has one statement?

Reply:
Probably. Not sure.